---
title: "specific_gene_expression"
author: "allyson_demerlis"
date: "2024-03-06"
output: html_document
---
# load packges and data
```{r}
library(tidyverse)
library(DESeq2)
library(cowplot)

load("Rdata_files/initial_fullddsdesigncountsVsdcounts.RData")
load("Rdata_files/realModels_Acer.RData")
load("Rdata_files/vsd.RData")
load("Rdata_files/pvals.RData")
```


# DEGs of interest
```{r}
Location_CCC_vs_Nursery_lfcshrink <- read_csv("results/resLFCLocation_CCC_vs_nursery_annotDGEs_padj05.csv")
Location_CCC_vs_Nursery_lfcshrink %>% 
  arrange(desc(log2FoldChange)) 

#Acropora_003744 = TNF associated factor
# Acropora_008410 = heat shock-mediated polytene chromosome puffing	
#Acropora_017542 = Belongs to the small heat shock protein (HSP20) family	
# Acropora_015083 = unfolded protein binding

Location_CCC_vs_Nursery_lfcshrink %>% 
  arrange(log2FoldChange)
#nothing sticks out to me 

design %>% 
  rownames_to_column(var="Sample") -> design
``` 


## Acropora_003744 = TNF associated factor
```{r}
p1<-t(vsd) %>% 
  as.data.frame() %>% 
  select(Acropora_003744) %>% 
  rownames_to_column() %>% 
  separate(rowname, into = c("Species", "ID", "Location", "Genotype"), sep ="_") %>% 
  unite("Sample", Species,ID, sep = "_") %>% 
  select(!Genotype) %>% 
  full_join(., design) %>% 
  ggplot(., aes(x=Location, y=Acropora_003744, fill = Location)) +
  geom_boxplot() +
  theme_classic() +
  labs(y = "VST-Normalized Expression Level", title = "Acropora_003744 = TNF associated factor") +
  theme(text = element_text(size = 15)) +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) 
```

## Acropora_008410 = heat shock-mediated polytene chromosome puffing	
```{r}
p2<-t(vsd) %>% 
  as.data.frame() %>% 
  select(Acropora_008410) %>% 
  rownames_to_column() %>% 
  separate(rowname, into = c("Species", "ID", "Location", "Genotype"), sep ="_") %>% 
  unite("Sample", Species,ID, sep = "_") %>% 
  select(!Genotype) %>% 
  full_join(., design) %>% 
  ggplot(., aes(x=Location, y=Acropora_008410, fill = Location)) +
  geom_boxplot() +
  theme_classic() +
  labs(y = "VST-Normalized Expression Level", title = "Acropora_008410 = heat shock-mediated polytene chromosome puffing") +
  theme(text = element_text(size = 15)) +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) 
```

## Acropora_017542 = Belongs to the small heat shock protein (HSP20) family	
```{r}
p3<-t(vsd) %>% 
  as.data.frame() %>% 
  select(Acropora_017542) %>% 
  rownames_to_column() %>% 
  separate(rowname, into = c("Species", "ID", "Location", "Genotype"), sep ="_") %>% 
  unite("Sample", Species,ID, sep = "_") %>% 
  select(!Genotype) %>% 
  full_join(., design) %>% 
  ggplot(., aes(x=Location, y=Acropora_017542, fill = Location)) +
  geom_boxplot() +
  theme_classic() +
  labs(y = "VST-Normalized Expression Level", title = "Acropora_017542 = Belongs to the small heat shock protein (HSP20) family") +
  theme(text = element_text(size = 15)) +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) 
```

## Acropora_015083 = unfolded protein binding
```{r}
p4<-t(vsd) %>% 
  as.data.frame() %>% 
  select(Acropora_015083) %>% 
  rownames_to_column() %>% 
  separate(rowname, into = c("Species", "ID", "Location", "Genotype"), sep ="_") %>% 
  unite("Sample", Species,ID, sep = "_") %>% 
  select(!Genotype) %>% 
  full_join(., design) %>% 
  ggplot(., aes(x=Location, y=Acropora_015083, fill = Location)) +
  geom_boxplot() +
  theme_classic() +
  labs(y = "VST-Normalized Expression Level", title = "Acropora_015083 = unfolded protein binding") +
  theme(text = element_text(size = 15)) +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) 
```


```{r}
cowplot::plot_grid(p1, p2, p3, p4, ncol = 1)
ggsave("specificgeneexpression.pdf", width = 12, height = 7)
```



# WGCNA genes

## darkmagenta module
```{r}
darkmagenta <- read_csv("WGCNA/darkmagenta_genelist.csv")

darkmagenta %>% 
  select(gene) %>% 
  distinct() -> darkmagentagenes

gene_annot <- read.table(file="bioinformatics/Acervicornis_iso2geneName.tab", sep = "\t", quote="", fill=FALSE)
gene_annot %>% 
  dplyr::rename(gene = V1, annot=V2) -> gene_annot

design %>% 
  rownames_to_column(var="Sample_ID") -> design

vsd %>% 
  as.data.frame() %>% 
  rownames_to_column(var="gene") %>% 
  arrange(desc(gene))

as.data.frame(resLFC) %>% 
  rownames_to_column(var="gene") -> allgenesl2fc

t(vsd) %>% 
  as.data.frame() %>% 
  rownames_to_column() %>% 
  separate(rowname, into = c("Species", "ID", "Location", "Genotype"), sep ="_") %>% 
  unite("Sample_ID", Species,ID, sep = "_") %>% 
  select(!Genotype) %>% 
  pivot_longer(Acropora_000001:Acropora_34621, names_to = "gene", values_to = "norm_expression") -> vst_normalized_counts

darkmagentagenes %>% 
  left_join(., gene_annot, by = "gene") %>% 
  drop_na(annot) %>% 
  left_join(., vst_normalized_counts) %>% 
  left_join(allgenesl2fc, by = "gene") %>% 
  left_join(., design, by = c("Sample_ID", "Location")) -> darkmagentagenes_annot_counts

#write_csv(darkmagentagenes_annot_counts, "darkmagentagenes_annot_counts_vstnormalized.csv")
```

### DEGs of interest in module
```{r}
darkmagentagenes_annot_counts %>% 
  arrange(desc(log2FoldChange))

# heat shock-mediated polytene chromosome puffing	
p1<-darkmagentagenes_annot_counts %>% 
  filter(annot == "heat shock-mediated polytene chromosome puffing" & gene == "Acropora_008410") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_008410: heat shock-mediated polytene chromosome puffing") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

# Belongs to the small heat shock protein (HSP20) family	
p2<-darkmagentagenes_annot_counts %>% 
  filter(annot == "Belongs to the small heat shock protein (HSP20) family" & gene == "Acropora_017542") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_017542: Belongs to the small heat shock protein (HSP20) family") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

# RAB30, member RAS oncogene family
p3<-darkmagentagenes_annot_counts %>% 
  filter(annot == "RAB30, member RAS oncogene family") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_009447: RAB30, member RAS oncogene family") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none") 

#G-protein coupled GABA receptor activity. It is involved in the biological process described with G-protein coupled receptor signaling pathway	
p4<-darkmagentagenes_annot_counts %>% 
  filter(annot == "G-protein coupled GABA receptor activity. It is involved in the biological process described with G-protein coupled receptor signaling pathway") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_026880: G-protein coupled GABA receptor activity") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

# heat shock
p5<-darkmagentagenes_annot_counts %>% 
  filter(annot == "heat shock") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_010472: heat shock") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

# Cytochrome P450, family 21, subfamily A, polypeptide 2
p6<-darkmagentagenes_annot_counts %>% 
  filter(annot == "Cytochrome P450, family 21, subfamily A, polypeptide 2") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_013946: Cytochrome P450") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

#	apoptotic process
p7<-darkmagentagenes_annot_counts %>% 
  filter(annot == "apoptotic process") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_010993: apoptotic process") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

plot_grid(p1,p2,p3,p4,p5,p6,p7,nrow=2)
#ggsave("darkmagentagenes.pdf", width = 12, height = 7)
```


### specific genes of interest
```{r}
mediumpurple3 <- read_csv("WGCNA/mediumpurple3_genelist.csv")

mediumpurple3 %>% 
  select(gene) %>% 
  distinct() -> mediumpurple3genes

mediumpurple3genes %>% 
  left_join(., gene_annot, by = "gene") %>% 
  drop_na(annot) %>% 
  left_join(., vst_normalized_counts) %>% 
  left_join(allgenesl2fc, by = "gene") %>% 
  left_join(., design, by = c("Sample_ID", "Location")) -> mediumpurple3genes_annot_counts

#write_csv(mediumpurple3genes_annot_counts, "mediumpurple3genes_annot_counts_vstnormalized.csv")

mediumpurple3genes_annot_counts %>% 
  arrange(desc(log2FoldChange))

# PIH1 domain-containing protein
m1<-mediumpurple3genes_annot_counts %>% 
  filter(annot == "PIH1 domain-containing protein") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_023565: PIH1 domain-containing protein") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

# assists the folding of proteins upon ATP hydrolysis
m2<-mediumpurple3genes_annot_counts %>% 
  filter(annot == "assists the folding of proteins upon ATP hydrolysis") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_015651: assists the folding of proteins upon ATP hydrolysis") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

# Matrix metalloproteinase-28
m3<-mediumpurple3genes_annot_counts %>% 
  filter(annot == "Matrix metalloproteinase-28") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_006280: Matrix metalloproteinase-28") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

# C-type lectin (CTL) or carbohydrate-recognition domain (CRD)
m4<-mediumpurple3genes_annot_counts %>% 
  filter(annot == "C-type lectin (CTL) or carbohydrate-recognition domain (CRD)") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_013375: C-type lectin (CTL) or carbohydrate-recognition domain (CRD)") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")


plot_grid(m1,m2,m3,m4,nrow=1)
#ggsave("mediumpurple3genes.pdf", width = 12, height = 7)
```



