---
title: "specific_gene_expression"
author: "allyson_demerlis"
date: "2024-03-06"
output: html_document
---
# load packges and data
```{r}
library(tidyverse)
library(DESeq2)
library(cowplot)

load("Rdata_files/initial_fullddsdesigncountsVsdcounts.RData")
load("Rdata_files/realModels_Acer.RData")

KEGG_genes <- read_csv("results/KEGG_CvN_results_withgenes.csv")
```

# normalize gene counts
```{r}
dds <- estimateSizeFactors(dds)
normalized_counts <- counts(dds, normalized=TRUE)

resLFC <- lfcShrink(dds, coef="Location_CCC_vs_nursery", type="apeglm")

normalized_counts %>% 
  as.data.frame() %>% 
  rownames_to_column(var="gene") -> normalized_counts

design %>% 
  rownames_to_column(var="Sample_ID") -> design
```

# join KEGG terms with gene counts
```{r}
KEGG_genes %>% 
  dplyr::select(ID, Description, Category, KO, gene, annot) %>% 
  left_join(., normalized_counts, by = "gene") %>%  #151 genes
  pivot_longer(Acer_1089:Acer_2384, names_to="Sample_ID", values_to="norm_expression") %>% 
  full_join(., design, by = "Sample_ID") -> KEGG_genes_normcounts


resLFC %>% 
  as.data.frame() %>% 
  rownames_to_column(var="gene") %>% 
  dplyr::select(gene, log2FoldChange) -> genes_l2fc

resLFC %>% 
  as.data.frame() %>% 
  rownames_to_column(var="gene") %>% 
filter(gene == "Acropora_013375")

KEGG_genes_normcounts %>% 
  right_join(., genes_l2fc, by="gene") -> KEGG_genes_normcounts_l2fc
```

Estrogen signaling pathway
```{r}
KEGG_genes_normcounts_l2fc %>% 
  filter(ID == "map04915") %>% 
  filter(norm_expression > 0) %>% 
  arrange(desc(log2FoldChange))
#heat shock-mediated polytene chromosome puffing	L2FC=6
# unfolded protein binding L2FC=1.95
# heat shock protein 90kDa beta (Grp94), member 1
# protein refolding

p1<-KEGG_genes_normcounts %>% 
  filter(ID == "map04915") %>% 
  filter(norm_expression > 0) %>% 
  filter(annot == "heat shock-mediated polytene chromosome puffing" & gene == "Acropora_008410") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  facet_wrap(~gene, scales = "free_y") +
  labs(y = "Normalized Expression Level", title = "map04915:Estrogen signaling pathway; Acropora_008410: heat shock-mediated polytene chromosome puffing") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

p2<-KEGG_genes_normcounts %>% 
  filter(ID == "map04915") %>% 
  filter(norm_expression > 0) %>% 
  filter(annot == "unfolded protein binding") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  facet_wrap(~gene, scales = "free_y") +
  labs(y = "Normalized Expression Level", title = "map04915:Estrogen signaling pathway; Acropora_017735: unfolded protein binding") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

p3<-KEGG_genes_normcounts %>% 
  filter(ID == "map04915") %>% 
  filter(norm_expression > 0) %>% 
  filter(annot == "heat shock protein 90kDa beta (Grp94), member 1") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  facet_wrap(~gene, scales = "free_y") +
  labs(y = "Normalized Expression Level", title = "map04915:Estrogen signaling pathway; Acropora_029829: heat shock protein 90kDa beta (Grp94), member 1") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

p4<-KEGG_genes_normcounts %>% 
  filter(ID == "map04915") %>% 
  filter(norm_expression > 0) %>% 
  filter(annot == "protein refolding") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  facet_wrap(~gene, scales = "free_y") +
  labs(y = "Normalized Expression Level", title = "map04915:Estrogen signaling pathway; Acropora_000806: protein refolding") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")
```

Melanogenesis
```{r}
p5<-KEGG_genes_normcounts_l2fc %>% 
  filter(ID == "map04916") %>% 
  filter(norm_expression > 0) %>% 
  arrange(desc(log2FoldChange)) %>% 
  filter(gene=="Acropora_015787") %>% #greatest L2FC
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot()+
  facet_wrap(~gene, scales = "free_y")+
  labs(y = "Normalized Expression Level", title = "map04916:Melanogenesis; Acropora_015787: calmodulin-dependent protein kinase activity") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")


#frizzled binding
p6<-KEGG_genes_normcounts %>% 
  filter(annot == "frizzled binding" & gene == "Acropora_009948") %>% 
  filter(norm_expression > 0) %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  facet_wrap(~gene, scales = "free_y")+
  labs(y = "Normalized Expression Level", title = "map04916:Melanogenesis; Acropora_009948: frizzled binding") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

```

Parathyroid hormone synthesis, secretion and action
```{r}
KEGG_genes_normcounts_l2fc %>% 
  filter(ID == "map04928") %>% 
  filter(norm_expression > 0) %>% 
  arrange(desc(log2FoldChange))
#cAMP response element binding	
# Matrix metalloproteinase-28	


p7<-KEGG_genes_normcounts %>% 
  filter(annot == "cAMP response element binding" & gene =="Acropora_000581") %>% 
  filter(norm_expression > 0) %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "map04928:Parathyroid hormone synthesis, secretion and action; Acropora_000581: cAMP response element binding") +
  facet_wrap(~gene, scales = "free_y")+
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

p8<-KEGG_genes_normcounts %>% 
  filter(annot == "Matrix metalloproteinase-28") %>% 
  filter(norm_expression > 0) %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  facet_wrap(~gene, scales = "free_y")+
  labs(y = "Normalized Expression Level", title = "map04928:Parathyroid hormone synthesis, secretion and action; Acropora_006280: Matrix metalloproteinase-28") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")
```

plot all genes together
```{r}
plot_grid(p1,p2,p3,p4,p5,p6,p7,p8, nrow=3)
#ggsave("kegg_genes_expression.pdf", width=12,height=7)
```

# WGCNA genes

## brown module
```{r}
brown2 <- read_csv("WGCNA/brown2_genelist.csv")

brown2 %>% 
  select(gene) %>% 
  distinct() -> browngenes

gene_annot <- read.table(file="bioinformatics/Acervicornis_iso2geneName.tab", sep = "\t", quote="", fill=FALSE)
gene_annot %>% 
  dplyr::rename(gene = V1, annot=V2) -> gene_annot

design %>% 
  rownames_to_column(var="Sample_ID") -> design

browngenes %>% 
  left_join(., gene_annot, by = "gene") %>% 
  drop_na(annot) %>% 
  left_join(., normalized_counts) %>% 
  pivot_longer(Acer_1089:Acer_2384, names_to="Sample_ID", values_to="norm_expression") %>% 
  left_join(., genes_l2fc) %>% 
  left_join(., design, by = "Sample_ID") -> browngenes_annot_counts

#write_csv(browngenes_annot_counts, "browngenes_annot_normcounts.csv")

# heat shock-mediated polytene chromosome puffing	
p1<-browngenes_annot_counts %>% 
  filter(annot == "heat shock-mediated polytene chromosome puffing" & gene == "Acropora_008410") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_008410: heat shock-mediated polytene chromosome puffing") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

# Belongs to the small heat shock protein (HSP20) family	
p2<-browngenes_annot_counts %>% 
  filter(annot == "Belongs to the small heat shock protein (HSP20) family" & gene == "Acropora_017542") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_017542: Belongs to the small heat shock protein (HSP20) family") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

# RAB30, member RAS oncogene family
p3<-browngenes_annot_counts %>% 
  filter(annot == "RAB30, member RAS oncogene family") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_009447: RAB30, member RAS oncogene family") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none") 

#G-protein coupled GABA receptor activity. It is involved in the biological process described with G-protein coupled receptor signaling pathway	
p4<-browngenes_annot_counts %>% 
  filter(annot == "G-protein coupled GABA receptor activity. It is involved in the biological process described with G-protein coupled receptor signaling pathway") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_026880: G-protein coupled GABA receptor activity") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

# heat shock
p5<-browngenes_annot_counts %>% 
  filter(annot == "heat shock") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_010472: heat shock") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

# Cytochrome P450, family 21, subfamily A, polypeptide 2
p6<-browngenes_annot_counts %>% 
  filter(annot == "Cytochrome P450, family 21, subfamily A, polypeptide 2") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_013946: Cytochrome P450") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

#	apoptotic process
p7<-browngenes_annot_counts %>% 
  filter(annot == "apoptotic process") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_010993: apoptotic process") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

plot_grid(p1,p2,p3,p4,p5,p6,p7,nrow=2)
#ggsave("brownmodule_genes.pdf", width = 12, height = 7)
```

## magenta module
```{r}
magenta3 <- read_csv("WGCNA/magenta3_genelist.csv")

magenta3 %>% 
  select(gene) %>% 
  distinct() -> magentagenes

magentagenes %>% 
  left_join(., gene_annot, by = "gene") %>% 
  drop_na(annot) %>% 
  left_join(., normalized_counts) %>% 
  pivot_longer(Acer_1089:Acer_2384, names_to="Sample_ID", values_to="norm_expression") %>% 
  left_join(., genes_l2fc) %>% 
  left_join(., design, by = "Sample_ID") -> magentagenes_annot_counts

#write_csv(magentagenes_annot_counts, "magenta3genes_annotated_normcounts.csv")

magentagenes_annot_counts %>% 
  arrange(log2FoldChange)

# PIH1 domain-containing protein
m1<-magentagenes_annot_counts %>% 
  filter(annot == "PIH1 domain-containing protein") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_023565: PIH1 domain-containing protein") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

# assists the folding of proteins upon ATP hydrolysis
m2<-magentagenes_annot_counts %>% 
  filter(annot == "assists the folding of proteins upon ATP hydrolysis") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_015651: assists the folding of proteins upon ATP hydrolysis") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

# Matrix metalloproteinase-28
m3<-magentagenes_annot_counts %>% 
  filter(annot == "Matrix metalloproteinase-28") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_006280: Matrix metalloproteinase-28") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

# C-type lectin (CTL) or carbohydrate-recognition domain (CRD)
m4<-magentagenes_annot_counts %>% 
  filter(annot == "C-type lectin (CTL) or carbohydrate-recognition domain (CRD)") %>% 
  ggplot(., aes(x=Location, y=norm_expression, fill=Location)) +
  geom_boxplot() +
  labs(y = "Normalized Expression Level", title = "Acropora_013375: C-type lectin (CTL) or carbohydrate-recognition domain (CRD)") +
  theme(text = element_text(size = 15)) +
  theme_classic() +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) +
  theme(legend.position = "none")

plot_grid(m1,m2,m3,m4,nrow=1)
#ggsave("magentagenes.pdf", width = 12, height = 7)
```




# DGEs of interest
```{r}
Location_CCC_vs_Nursery_lfcshrink %>% 
  arrange(desc(log2FoldChange)) 

#Acropora_003744 = TNF associated factor
# Acropora_008410 = heat shock-mediated polytene chromosome puffing	
#Acropora_017542 = Belongs to the small heat shock protein (HSP20) family	
# Acropora_015083 = unfolded protein binding
#Acropora_011526 = T cell activation via T cell receptor contact with antigen bound to MHC molecule on antigen presenting cell	

Location_CCC_vs_Nursery_lfcshrink %>% 
  arrange(log2FoldChange)
#nothing sticks out to me 
``` 

# normalize gene counts
```{r}
dds <- estimateSizeFactors(dds)
normalized_counts <- counts(dds, normalized=TRUE)

# Subset normalized counts based on a criterion, e.g., adjusted p-value less than 0.05
# First, identify which genes meet the criterion
significant_genes <- rownames(results[!is.na(results$padj) & results$padj < 0.05, ])

# Then, subset the normalized counts based on these genes
significant_counts <- normalized_counts[significant_genes, ]
```


# boxplot Acropora_003744 = TNF associated factor
```{r}
Acropora_003744_counts <- significant_counts["Acropora_003744",]

df_Acropora_003744 <- data.frame(expression_level = Acropora_003744_counts)

df_Acropora_003744 %>% 
  rownames_to_column(var="Sample") -> df_Acropora_003744

design %>% 
  rownames_to_column(var="Sample") -> design
  
p1<-full_join(design, df_Acropora_003744) %>% 
  ggplot(., aes(x=Location, y=expression_level, fill=Location)) +
  geom_boxplot() +
  theme_classic() +
  labs(y = "Normalized Expression Level", title = "Acropora_003744 = TNF associated factor") +
  theme(text = element_text(size = 15)) +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) 
```

# boxplot of Acropora_008410 = heat shock-mediated polytene chromosome puffing	
```{r}
Acropora_008410_counts <- normalized_counts["Acropora_008410",]

df_Acropora_008410 <- data.frame(expression_level = Acropora_008410_counts)

df_Acropora_008410 %>% 
  rownames_to_column(var="Sample") -> df_Acropora_008410
 
p2<-full_join(design, df_Acropora_008410) %>% 
  ggplot(., aes(x=Location, y=expression_level, fill=Location)) +
  geom_boxplot() +
  theme_classic() +
  labs(y = "Normalized Expression Level", title = "Acropora_008410 = heat shock-mediated polytene chromosome puffing	") +
  theme(text = element_text(size = 15)) +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) 
```

#Acropora_017542 = Belongs to the small heat shock protein (HSP20) family	
```{r}
Acropora_017542_counts <- normalized_counts["Acropora_017542",]

df_Acropora_017542 <- data.frame(expression_level = Acropora_017542_counts)

df_Acropora_017542 %>% 
  rownames_to_column(var="Sample") -> df_Acropora_017542
 
p3<-full_join(design, df_Acropora_017542) %>% 
  ggplot(., aes(x=Location, y=expression_level, fill=Location)) +
  geom_boxplot() +
  theme_classic() +
  labs(y = "Normalized Expression Level", title = "Acropora_017542 = Belongs to the small heat shock protein (HSP20) family") +
  theme(text = element_text(size = 15)) +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) 
```

# Acropora_015083 = unfolded protein binding
```{r}
Acropora_015083_counts <- normalized_counts["Acropora_015083",]

df_Acropora_015083 <- data.frame(expression_level = Acropora_015083_counts)

df_Acropora_015083 %>% 
  rownames_to_column(var="Sample") -> df_Acropora_015083
 
p4<-full_join(design, df_Acropora_015083) %>% 
  ggplot(., aes(x=Location, y=expression_level, fill=Location)) +
  geom_boxplot() +
  theme_classic() +
  labs(y = "Normalized Expression Level", title = "Acropora_015083 = unfolded protein binding") +
  theme(text = element_text(size = 15)) +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) 
```

# Acropora_011526 = T cell activation via T cell receptor contact with antigen bound to MHC molecule on antigen presenting cell	
```{r}
Acropora_011526_counts <- normalized_counts["Acropora_011526",]

df_Acropora_011526 <- data.frame(expression_level = Acropora_011526_counts)

df_Acropora_011526 %>% 
  rownames_to_column(var="Sample") -> df_Acropora_011526
 
p5<-full_join(design, df_Acropora_011526) %>% 
  ggplot(., aes(x=Location, y=expression_level, fill=Location)) +
  geom_boxplot() +
  theme_classic() +
  labs(y = "Normalized Expression Level", title = "Acropora_011526 = T cell activation ") +
  theme(text = element_text(size = 15)) +
  scale_fill_manual(values = c("CCC" = "orange", "nursery" = "darkblue")) 
```


```{r}
cowplot::plot_grid(p1, p2, p3, p4, p5, ncol = 3)
ggsave("specificgeneexpression.pdf", width = 12, height = 7)
```

