---
title: "AcerCCC_readcounts_updatedgff3"
author: "allyson_demerlis"
date: "2023-07-21"
output: html_document
---

```{r}
library(data.table)
library(ggplot2)
library(ggrepel)
library(dplyr)
library(tidyr)
library(reshape2)
library(tidyverse)
```


```{r}
#Sorting samples with less than 4M and 5M reads mapped

reads_mapped <- read.delim("../bioinformatics/multiqc_reports/multiqc_general_stats_STARalign_trimmed_updatedgfftake3.txt", sep = "\t")
head(reads_mapped)

reads_mapped<- reads_mapped %>% 
           dplyr::rename(percent_uniq = STAR_mqc.generalstats.star.uniquely_mapped_percent) %>%
  
               dplyr::rename(million_uniq = STAR_mqc.generalstats.star.uniquely_mapped) %>% 
  
               dplyr::rename(Sample_ID = Sample)

samples_selected_4M<- reads_mapped %>% filter(million_uniq > 4000000)
list_sampleIDs <- samples_selected_4M %>% select(Sample_ID) %>% as.list()

samples_selected_5M<- reads_mapped %>% filter(million_uniq > 5000000)
```


```{r}
# read in metadata
sample_metadata <- read.csv("../input_files/sample_metadata.csv")
names(sample_metadata)

#join read counts
sample_metadata_reads<- sample_metadata %>% left_join(samples_selected_4M, by="Sample_ID") 

sample_metadata_reads <- drop_na(sample_metadata_reads)
```


```{r}
# create a list of all files from samples that have at least 4M reads mapped

file_4M<- list.files("../star_trimmed_readsPerGeneoutTab_updatedgff3",
                  "*.gzReadsPerGene.out.tab", full.names = T)

countData_4M = data.frame(fread(file_4M[1]))[c(1,3)]

for(i in 2:length(file_4M)) {
        countData_4M = cbind(countData_4M, data.frame(fread(file_4M[i]))[3])
}

# Skip first 4 lines, count data starts on the 5th line
countData_4M = countData_4M[c(5:nrow(countData_4M)),]
colnames(countData_4M) = c("GeneID", gsub(paste0("../star_trimmed_readsPerGeneoutTab_updatedgff3"), "", file_4M))
colnames(countData_4M) = gsub("_trimmed_trimmed.fastq.gzReadsPerGene.out.tab", "",colnames(countData_4M))
colnames(countData_4M) = gsub("/", "ID_",colnames(countData_4M))
rownames(countData_4M) = countData_4M$GeneID
str(countData_4M) #33715 x 18

countData_4M = countData_4M[,c(2:ncol(countData_4M))] #this gets rid of the GeneID column but keeps GeneIDs as rownames 
str(countData_4M) #33715 x 17

#write_rds(countData_4M, "countData_4M.rds")
```


Plot read counts
```{r}
ggplot(samples_selected_4M, aes(million_uniq)) + geom_histogram()

ggplot(samples_selected_4M, aes(percent_uniq)) + geom_histogram()

sample_metadata_reads %>% filter(!Genotype=="stag hybrid") %>% 
  ggplot(.) + 
  geom_bar(aes(x=Sample_ID, y=million_uniq, fill=Genotype), stat = "identity", position = position_dodge())

sample_metadata_reads %>% filter(!Genotype=="stag hybrid") %>% 
  ggplot(.) + 
  geom_bar(aes(x=Sample_ID, y=million_uniq, fill=Genotype), stat = "identity", position = position_dodge()) + 
  facet_wrap(~Location, scales = "free_x") +
  theme_classic() +
    theme(axis.text.x=element_text(angle=45,hjust=1))
```

